Assets/Scripts/SS3D/Systems/Substances/SubstanceContainerActor.cs

```diff
@@ -0,0 +1,618 @@
+        {
+            substanceContainer.RemoveSubstance(substance, moles);
+        }
+
+        /// <summary>
+        /// Empties the container
+        /// </summary>
+        public void Empty()
+        {
+            substanceContainer.Empty();
+        }
+
+        /// <summary>
+        /// Removes moles from the container
+        /// </summary>
+        /// <param name="moles">The amount of moles</param>
+        [Server]
+        public void RemoveMoles(float moles)
+        {
+            substanceContainer.RemoveMoles(moles);
+        }
+
+        /// <summary>
+        /// Transfers moles to a different container
+        /// </summary>
+        /// <param name="other">The other container</param>
+        /// <param name="moles">The moles to transfer</param>
+        [Server]
+        public void TransferMoles(SubstanceContainerActor other, float moles)
+        {
+            substanceContainer.TransferMoles(other.substanceContainer, moles);
+        }
+
+        /// <summary>
+        /// Transfers volume (ml) to a different container
+        /// </summary>
+        /// <param name="other">The other container</param>
+        /// <param name="milliliters">How many milliliters to transfer</param>
+        [Server]
+        public void TransferVolume(SubstanceContainerActor other, float milliliters)
+        {
+            substanceContainer.TransferVolume(other.substanceContainer, milliliters);
+        }
+
+        /// <summary>
+        /// Returns the index of an substance
+        /// </summary>
+        /// <param name="substance">The substance to look for</param>
+        /// <returns>The index or -1 if it is not found</returns>
+        [Server]
+        public int IndexOfSubstance(Substance substance)
+        {
+            return substanceContainer.IndexOfSubstance(substance);
+        }
+
+        /// <summary>
+        /// Informs the system that the contents of this container have changed
+        /// </summary>
+        [Server]
+        public void MarkDirty()
+        {
+            OnContentsChanged();
+        }
+
+        [Server]
+        protected virtual void OnContentsChanged()
+        {
+             ProcessContainer(this);
+            ContentsChanged?.Invoke(this);
+        }
+
+        public override IInteraction[] CreateTargetInteractions(InteractionEvent interactionEvent)
+        {
+            return new IInteraction[]
+            {
+                new TransferSubstanceInteraction()
+            };
+        }
+
+        /// <summary>
+        /// Processes substances in a container
+        /// </summary>
+        public static void ProcessContainer(SubstanceContainerActor container)
+        {
+            SubstanceContainer.ProcessContainer(container.substanceContainer);
+            container.MarkDirty();
+        }
+
+        #endregion
+
+        #region SubstanceContainer
+        public class SubstanceContainer
```

Comment:
Is the SubstanceContainer class being inside the SubstanceContainerActor mandatory? The style guide enforces only one class per file

Reply:
Sending you what I said to John on Discord about that : 

> So you know how we talk about the humble object pattern ? Having network logic in one class and another class handling all the system, object whatever logic ? Well, after scratching my head for a while, I found out having a nested class for the humble object is just the best way to preserve encapsulation in the "network-handling" class.

There might be a better way, but I believe this is a case where it actually make sense. I got a few syncvars in substanceContainerActor, and I don't want them to be public in any way, but I need substanceContainer to still act on them, and I can't move them to SubstanceContainer because they're syncvars. I'd be happy if there's a better way to have the syncvar private and still act on them in SubstanceContainer only, sadly I don't think C# has any equivalent of Friend in C++. Another way would be to have public internal and put those two files in their own assembly..